python/tank/util/shotgun/publish_resolve.py

```diff
@@ -0,0 +1,469 @@
+              a :class:`~sgtk.util.PublishPathNotDefinedError` is raised.
+
+    **Resolving file urls**
+
+    The method also supports the resolution of ``file://`` urls. Such urls
+    are not multi platform and the local storages and environment variables
+    will therefore be used to try to resolve such paths in case of ambiguity.
+
+    - First, local storage settings will be downloaded from Shotgun and added to
+      a path translation table.
+
+    - Next, similar to the local file linking above, any environment variables
+      on the form ``SHOTGUN_PATH_WINDOWS|MAC|LINUX`` or ``SHOTGUN_PATH_WINDOWS|MAC|LINUX_STORAGENAME``
+      will be added to the translation table.
+
+    - The ``file://`` path will be resolved compared against all the existing roots
+      for all operating systems. The first match detected will be used to translate
+      the path into the current operating system platform.
+
+    - If there is no against any storage, the file path be returned.
+
+    For example, you have published the file ``/projects/some/file.txt`` on Linux
+    and generated a publish with the url ``file:///projects/some/file.txt``. You have
+    either a local storage or environment variable set up with the following paths:
+
+    - Linux Path: ``/projects``
+    - Windows Path: ``Q:\projects``
+    - Mac Path: ``/projects``
+
+    When running on windows, the ``file://`` url will therefore be translated to
+    ``Q:\\projects\\some\\file.txt``.
+
+    .. note:: If no value has been defined for the current operating system,
+              a :class:`~sgtk.util.PublishPathNotDefinedError` is raised.
+
+    **Customization examples**
+
+    If you want to add support beyond local file links and ``file://`` urls, you can
+    customize the ``resolve_publish.py`` core hook. This can for example be used to
+    add support for the following customizations:
+
+    - Publishes with associated uploaded files could be automatically downloaded
+      into an appropriate cache location by the core hook and the path would be
+      be returned.
+
+    - Custom url schemes (such as ``perforce://``) could be resolved into local paths.
+
+
+    **Parameters**
+
+    :param tk: :class:`~sgtk.Sgtk` instance
+    :param sg_publish_data: Dictionary containing Shotgun publish data.
+        Needs to at least contain a code, type, id and a path key.
+
+    :returns: A local path to file or file sequence.
+
+    :raises: :class:`~sgtk.util.PublishPathNotDefinedError` if the path isn't defined.
+    :raises: :class:`~sgtk.util.PublishPathNotSupported` if the path cannot be resolved.
+    """
+
+    path_field = sg_publish_data.get("path")
+
+    log.debug(
+        "Publish id %s: Attempting to resolve publish path "
+        "to local file on disk: '%s'" % (sg_publish_data["id"], pprint.pformat(path_field))
+    )
+
+    # first offer it to the core hook
+    custom_path = tk.execute_core_hook_method(
```

Comment:
yes, interestingly i gave this extensive thought! I will add more to comments. great feedback.

Basically, we want to be able to add support for handling uploaded files (or something else) in the future, yet at the same time, by doing that we don't want to break any client integration. By putting the hook first, this is possible. If we put the hook last, we could affect the conditions under which the hooks is being executed by introducing new features.